### PR TITLE
[#9255] Migrate instructor course edit page to Angular

### PR DESCRIPTION
**Pull request by joanneong**
Part of #9255

v7 look:

image

v6 look:

image

Summary of what has been done:
- Migrated actions include getting course details, editing course details, deleting course, adding instructor, editing instructor, deleting instructor and sending reminder emails
  - Each action involve a new api endpoint, with the exception of deleting instructor which modified `DeleteInstructorAction`
- Migrated front-end ~(incomplete)~
- Migrated tests for most actions except for deleting instructor, which has to be written from scratch (it seems)

Some observations so far:
- Some visible lag time taken to load the page
- Timezones are missing the exact UTC offset as the `TimezoneService` does not seem to provide it yet
- ~Auto-detect button is not functional - again, `TimezoneService` does not seem to support this~
- Change in position for status messages in course edit page (used to be in between the course details panel and the instructor details panels)
- Code has not been refactored - html can be repetitive for instructors panel (can consider extracting into components in the future) - for Modals as well
- Horrible way of circumventing non-null checks since using `!` will fail `npm run lint`
- When you refresh the page, the instructor panels are reordered by google id in alphabetical order
- Because I was using reactive forms in the html for the page, I decided to avoid using template syntax as far as possible so two-way binding may appear extraordinarily lengthy. Not too sure if this is a good way to go about doing things

To do:
- [x] Write `DeleteInstructorActionTest`
- [x] Route links correctly (e.g. `View Details`)
- [x] Create modals for instructor privileges
- [x] Edit style for displaying google id for instructors when there is no id available yet
- [x] Add button to resend invitation email for a newly added instructor
- [ ] Check that no interactions have been missed out
- [ ] Delete old files after migration

### CODE REVIEW

**Comment by wkurniawan07** [comment]
> * Auto-detect button is not functional - again, `TimezoneService` does not seem to support this

Just use `moment` library? This is already done at least in admin sessions page.

**Comment by joanneong** [comment]
I actually referenced that page at first to find the `TimezoneService`, bu

**Comment by joanneong** [comment]
Progress update: I have missed out a panel for when the `Custom` role is selected. Other than that, the rest of the interactions should be completed (but will still need to check through all the files to confirm).

**Comment by wkurniawan07** [comment]
@joanneong you need to resolve the merge conflicts.

File: src/web/app/pages-instructor/instructor-course-edit-page/instructor-course-edit-page.component.ts [hunk 1]
```diff
@@ -12,13 +71,627 @@ import { ActivatedRoute } from '@angular/router';
...
     });
   }
 
+  /**
+   * Adds a new instructor to the current course.
+   */
+  onSubmitAddInstructor(formAddInstructor: FormGroup): void {
+    // Create a copy of the added instructor
+    const addedInstructor: InstructorAttributes = {
+      googleId: formAddInstructor.controls.googleId.value,
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-course-edit-page/instructor-course-edit-page.component.ts]: This is causing bug that prevents instructor from being added. When adding a new instructor, they would not have a google ID yet.

File: src/web/app/pages-instructor/instructor-course-edit-page/instructor-course-edit-page.component.html [hunk 2]
```diff
@@ -1,3 +1,368 @@
...
+            </div>
+            <div>
+              <div class="form-row form-group">
+                <label class="col-sm-3 control-label">
+                  Access Level:
+                </label>
+                <div class="col-sm-9">
+                  <span class="form-control-static" id="role-view-{{ i }}">
+                    {{ instructorList[i].role }}
+                    <a class="view-role-details" href="#" (click)="viewInstructorRole(viewInstructorRoleModal, i)">
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-course-edit-page/instructor-course-edit-page.component.html]: `href="#"` will bring you to the top of the page. You're likely looking for `href="javascript:;"`?

File: src/web/app/pages-instructor/instructor-course-edit-page/instructor-course-edit-page.component.ts [hunk 3]
```diff
@@ -12,13 +71,627 @@ import { ActivatedRoute } from '@angular/router';
...
+    }
+  }
+
+  /**
+   * Deletes the current course and redirects to 'Courses' page if action is successful.
+   */
+  deleteCourse(): void {
+    const paramsMap: { [key: string]: string } = {
+      courseid: this.courseToEdit.id,
+      next: '/web/instructor/courses',
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-course-edit-page/instructor-course-edit-page.component.ts]: There shouldn't be a need to pass this to the back-end.

**Comment by joanneong** [comment]
Updates:
- Being able to access a soft-deleted course is actually a bug carried over from v6 (you can access with the same link as well). I have fixed this by checking that the course is not soft-deleted before displaying the page. I have also added a test to test for this situation.
- Tooltip for the `Delete` button for an instructor panel is sometimes unstable (?)
- Still looking for a better solution to the DOM manipulation
- Still missing custom privileges

**Pull request merged**